### PR TITLE
[sw/otbn/crypto] Add arithmetic to boolean masking for P256 ECDH

### PR DESCRIPTION
**Pull request by wettermo**
This PR/commit modifies the p256_scalar_mult function so that it returns an arithmetically masked result. The arithmetic masking is done before converting back to affine coordinates. For this the internal scalar_mult_int function is changed to return projective coordinates. The transformation to affine coordinates now needs to be done outside scalar_mult_int.

This commit also adds code to convert from arithmetic masking to boolean masking. The conversion is implemented following Gaubin's algorithm to prevent side-channel leakage. The conversion is implemented in a 257-bit domain to enable carry bit handling, which is necessary when the arithmetic masking is performed within a mod p domain.

Both described changes are then put together to generate the P256 ECDH shared session key.

Further, this commit also adds the relevant tests.

UPDATE:
This PR also restructures the code to prevent an overflow of OTBN IMEM (see comment further below).

### CODE REVIEW

File: sw/otbn/crypto/p256.s [hunk 1]
```diff
@@ -2254,6 +2325,256 @@ boolean_to_arithmetic:
...
+     w20 <= w20 */
+  bn.mov    w20, w20
+
+  ret
+
+  /* Return result of reduced A2B computation (first result). */
+  arithmetic_to_boolean_mod_ret_reduced:
+  /* Restore previous result to output WDR.
+     w20 <= w26 */
+  bn.mov    w20, w26
```
Comment (jadephilipoom) [inline sw/otbn/crypto/p256.s:2458]: I think you can do this whole operation  with `bn.sel`:
```suggestion
bn.sel    w20, w20, w26, FG0.Z
```

**Review by jadephilipoom — APPROVED** [review]

File: sw/otbn/crypto/p256.s [hunk 2]
```diff
@@ -2254,6 +2325,256 @@ boolean_to_arithmetic:
 
   ret
 
+/**
+ * Converts arithmetic shares within the "mod p" scalar field to
+ * boolean shares by calling the 257-bit A2B function twice,
+ * first using unmodified 256-bit shares in reduced form,
+ * and then using modified 257-bit shares in unreduced form.
```
[the remark this answers is not in the document]
Comment (wettermo) [reply sw/otbn/crypto/p256.s]: Definitely makes sense to begin with a short description. Thanks for pointing it out!

File: sw/otbn/crypto/p256.s [hunk 3]
```diff
@@ -2254,6 +2325,256 @@ boolean_to_arithmetic:
...
+ * (2^257 - p + x - r + r) mod 2^257 = 2^257 - p + x
+ * (high bit is always true since p - x <= p < 2^256) for the x >= r case,
+ * or (2^257 - p + x - r + p + r) mod 2^257 = (2^257 + x) mod 2^257 = x
+ * (high bit is always false since x < p < 2^256) for the r > x case.
+ * So when the high bit is false, we pick this version, with 2^257 - p added.
+ * But if the high bit is true, then when we take the earlier computed
+ * boolean shares for (A + r) mod 2^257, as they would have been valid
+ * boolean shares for x since A = x - r exactly, no reduction needed,
+ * and therefore A + r = x. So if the high bit is true,
+ * we select the version without the 2^257 - p factor added.
```
[the remark this answers is not in the document]
Comment (wettermo) [reply sw/otbn/crypto/p256.s]: Thanks a lot, changed it.

File: sw/otbn/crypto/p256.s [hunk 4]
```diff
@@ -2254,6 +2325,256 @@ boolean_to_arithmetic:
...
+     w11 <= w25
+     w12 <= w31 */
+  bn.mov    w18, w24
+  bn.mov    w19, w31
+  bn.mov    w11, w25
+  bn.mov    w12, w31
+
+  /* Get field modulus p.
+     w29 <= MOD() */
+  bn.wsrr   w29, 0x00 /* MOD */
```
[the remark this answers is not in the document]
Comment (wettermo) [reply sw/otbn/crypto/p256.s:2408]: Noted :)

File: sw/otbn/crypto/p256.s (hunk 1 above)
Comment (wettermo) [reply sw/otbn/crypto/p256.s]: That's actually a very good improvement. I changed it and it still works as intended. Thank you!

**Comment by wettermo** [comment]
I just saw that with this code the device driver for ECDH [ecdh_p256.c](https://github.com/lowRISC/opentitan/blob/master/sw/device/lib/crypto/impl/ecc/ecdh_p256.c) needs to be adapted. The reason is that now the second share of the shared key is returned in DMEM via the variable `m_x` (before it was returned via `y`). 
Probably the best way to solve this is to return the second share again via `y`, this way the device driver doesn't have to be changed. What do you think @jadephilipoom ?

**Comment by jadephilipoom** [comment]
> I just saw that with this code the device driver for ECDH [ecdh_p256.c](https://github.com/lowRISC/opentitan/blob/master/sw/device/lib/crypto/impl/ecc/ecdh_p256.c) needs to be adapted. The reason is that now the second share of the shared key is returned in DMEM via the variable `m_x` (before it was returned via `y`). Probably the best way to solve this is to return the second share again via `y`, this way the device driver doesn't have to be changed. What do you think @jadephilipoom ?

I think the changes to the device driver should be pretty minimal, just editing the declaration of the symbol and then the variable name that gets used later -- maybe it would be nice to change it since `m_x` makes a little more sense in this context.

I think all that needs to change is these two lines:
https://github.com/lowRISC/opentitan/blob/dd8b7098f35b43190506a33460b2852b9edf6155/sw/device/lib/crypto/impl/ecc/ecdh_p256.c#L19
https://github.com/lowRISC/opentitan/blob/dd8b7098f35b43190506a33460b2852b9edf6155/sw/device/lib/crypto/impl/ecc/ecdh_p256.c#L28
...and then changing the variable name `kOtbnVarEcdhY` to `kOtbnVarEcdhMx` for style/clarity.

**Comment by wettermo** [comment]
Okay I'll change it in a separate PR.

But I would rather add `kOtbnVarEcdhMx` instead of replacing `kOtbnVarEcdhY` with it, because `kOtbnVarEcdhY` is still used as input for the public key y-coordinate. Then only `ecdh_p256_shared_key_finalize()` needs to be changed.

**Comment by jadephilipoom** [comment]
Oh, I see, in that case it actually does make sense to change `m_x` to `y` so we can reuse the buffer! I thought it was only an output buffer but had a mismatched name. I retract my previous comment, no device driver changes needed :slightly_smiling_face:

**Comment by wettermo** [comment]
Okay, but for clearness, this would mean that already `p256_scalar_mult` in [p256.s](https://github.com/lowRISC/opentitan/blob/master/sw/otbn/crypto/p256.s) wouldn't return a y-coordinate anymore, only the masked x-coordinate as `x` and the mask as `y`.

**Comment by jadephilipoom** [comment]
> Okay, but for clarity, this would mean that already `p256_scalar_mult` in [p256.s](https://github.com/lowRISC/opentitan/blob/master/sw/otbn/crypto/p256.s) wouldn't return a y-coordinate anymore, only the masked x-coordinate as `x` and the mask as `y`.

I think that's okay. We should only need `p256_scalar_mult` for ECDH, since we use `p256_base_mult` instead to generate the public key, and ECDSA uses the internal scalar mult.

**Comment by wettermo** [comment]
I changed the code so that `y` is again used as return value for the mask. This means there is no need to change the device driver. I also changed and tested the tests accordingly.

**Review by jadephilipoom — APPROVED** [review]

**Comment by wettermo** [comment]
last force-push fixed some comments: because of this PR#s changes output parameters of `scalar_mult_int` were commented incorrectly

**Comment by wettermo** [comment]
The latest added (second) commit restructures the code to prevent an
overflow of OTBN IMEM.

The changes to P256 OTBN software, introduced in the previous commit,
increase code size so much that IMEM of OTBN is too small for the
application. Therefore, this commit restructures P256 code to prevent
an IMEM overflow. For this, p256.s is split into several individual
files containing different functions. In case the functions are not
needed for a specific binary, they now won't get inserted into the
binary anymore.

**Review by jadephilipoom — APPROVED** [review]

**Pull request merged**